[Site Isolation] Fix Spell Check

stwrt opened the pull request:
#### c052807daec1ac946eb59c98dc6d5c261e5382f7
<pre>
[Site Isolation] Fix Spell Check
<a href="https://bugs.webkit.org/show_bug.cgi?id=292012">https://bugs.webkit.org/show_bug.cgi?id=292012</a>
<a href="https://rdar.apple.com/117912363">rdar://117912363</a>

Reviewed by Alex Christensen.

When calculating the bounding box, the inner iframe is not able to see the outer iframes location. To correct for this missing
information, the UI process will then correct the received bounding boxes location to draw the correct UI pop up.

* Source/WebCore/editing/AlternativeTextController.cpp:
(WebCore::AlternativeTextController::show):
(WebCore::AlternativeTextController::timerFired):
* Source/WebCore/page/AlternativeTextClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::showCorrectionPanel):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebAlternativeTextClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebAlternativeTextClient.cpp:
(WebKit::WebAlternativeTextClient::showCorrectionAlternative):
* Source/WebKitLegacy/mac/WebCoreSupport/WebAlternativeTextClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebAlternativeTextClient.mm:
(WebAlternativeTextClient::showCorrectionAlternative):

Canonical link: <a href="https://commits.webkit.org/294111@main">https://commits.webkit.org/294111@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/24d1ef2a75a6a17e5683eb51b43d0e2407c0c77e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100832 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20484 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10783 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105969 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51421 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102873 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20793 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/29006 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76777 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33815 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103839 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15982 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91075 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57129 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15793 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9084 "Found 6 new test failures: fonts/monospace.html fonts/sans-serif.html fonts/serif.html imported/w3c/web-platform-tests/svg/painting/marker-005.svg imported/w3c/web-platform-tests/svg/painting/marker-006.svg media/media-source/media-source-video-renders.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50797 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85693 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9158 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108324 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27950 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20533 "Build was cancelled. Recent messages:OS: Sequoia (15.3), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 10 flakes 8 failures; Uploaded test results; Exiting early after 60 failures. 76947 tests run. 60 failures; Compiled WebKit (warnings); layout-tests (cancelled)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85738 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28313 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87277 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85281 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21712 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29994 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7724 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21959 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27885 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33142 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27696 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31014 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29254 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKitLegacy/mac/WebCoreSupport/WebAlternativeTextClient.mm:
```diff
@@ -45,8 +45,9 @@
-void WebAlternativeTextClient::showCorrectionAlternative(AlternativeTextType type, const FloatRect& boundingBoxOfReplacedString, const String& replacedString, const String& replacementString, const Vector<String>& alternativeReplacementStrings)
+void WebAlternativeTextClient::showCorrectionAlternative(AlternativeTextType type, const FloatRect& boundingBoxOfReplacedString, const String& replacedString, const String& replacementString, const Vector<String>& alternativeReplacementStrings, std::optional<FrameIdentifier> rootFrame)
 {
+    UNUSED_PARAM(rootFrame);
```
Just omit the parameter name.
..., FrameIdentifier)

achristensen07 commented on Source/WebKit/WebProcess/WebCoreSupport/mac/WebAlternativeTextClient.cpp:
```diff
@@ -47,9 +47,10 @@ WebAlternativeTextClient::~WebAlternativeTextClient()
+void WebAlternativeTextClient::showCorrectionAlternative(AlternativeTextType type, const FloatRect& boundingBoxOfReplacedString, const String& replacedString, const String& replacementString, const Vector<String>& alternativeReplacementStrings, std::optional<FrameIdentifier> rootFrame)
 {
-    m_page->send(Messages::WebPageProxy::ShowCorrectionPanel(type, boundingBoxOfReplacedString, replacedString, replacementString, alternativeReplacementStrings));
+    UNUSED_PARAM(rootFrame);
```
This is not needed and incorrect.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.messages.in:
```diff
@@ -379,7 +379,7 @@ messages -> WebPageProxy {
 #if PLATFORM(MAC)
     # Autocorrection messages
-    ShowCorrectionPanel(enum:uint8_t WebCore::AlternativeTextType panelType, WebCore::FloatRect boundingBoxOfReplacedString, String replacedString, String replacementString, Vector<String> alternativeReplacementStrings)
+    ShowCorrectionPanel(enum:uint8_t WebCore::AlternativeTextType panelType, WebCore::FloatRect boundingBoxOfReplacedString, String replacedString, String replacementString, Vector<String> alternativeReplacementStrings, std::optional<WebCore::FrameIdentifier> rootFrameID)
```
I think this parameter shouldn't be optional.  If we don't have a root frame ID, we are disconnected from the frame tree and should early return.

stwrt replied:
I designed it to use an optional based on `convertRectToMainFrameCoordinates`, which take an optional frame identifier.

```
void WebPageProxy::convertRectToMainFrameCoordinates(WebCore::FloatRect rect, std::optional<WebCore::FrameIdentifier> frameID, CompletionHandler<void(std::optional<WebCore::FloatRect>)>&& completionHandler)
```

achristensen07 replied:
At this point, though, it shouldn't be optional.  Same with AlternativeTextClient.  If we don't have a root frame ID, we shouldn't be making the AlternativeTextClient call.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -13139,10 +13139,14 @@ void WebPageProxy::substitutionsPanelIsShowing(CompletionHandler<void(bool)>&& c
 }
 
-void WebPageProxy::showCorrectionPanel(AlternativeTextType panelType, const FloatRect& boundingBoxOfReplacedString, const String& replacedString, const String& replacementString, const Vector<String>& alternativeReplacementStrings)
+void WebPageProxy::showCorrectionPanel(AlternativeTextType panelType, const FloatRect& boundingBoxOfReplacedString, const String& replacedString, const String& replacementString, const Vector<String>& alternativeReplacementStrings, std::optional<FrameIdentifier> rootFrameID)
```
Change alternativeReplacementStrings from a const Vector<String>& to a Vector<String>&&.  Same with the const String& parameters.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -13139,10 +13139,14 @@ void WebPageProxy::substitutionsPanelIsShowing(CompletionHandler<void(bool)>&& c
 {
-    if (RefPtr pageClient = this->pageClient())
-        pageClient->showCorrectionPanel(panelType, boundingBoxOfReplacedString, replacedString, replacementString, alternativeReplacementStrings);
+    if (RefPtr pageClient = this->pageClient()) {
```
nit: an early return if pageClient is null would reduce indentation a bit.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -13139,10 +13139,14 @@ void WebPageProxy::substitutionsPanelIsShowing(CompletionHandler<void(bool)>&& c
+    if (RefPtr pageClient = this->pageClient()) {
+        convertRectToMainFrameCoordinates(boundingBoxOfReplacedString, rootFrameID, [pageClient = WTFMove(pageClient), panelType = WTFMove(panelType), boundingBoxOfReplacedString = WTFMove(boundingBoxOfReplacedString), replacedString = WTFMove(replacedString),
+            replacementString = WTFMove(replacementString), alternativeReplacementStrings = WTFMove(alternativeReplacementStrings)](std::optional<WebCore::FloatRect> correctedBoundingBox) {
+            pageClient->showCorrectionPanel(panelType, correctedBoundingBox ? correctedBoundingBox.value() : boundingBoxOfReplacedString, replacedString, replacementString, alternativeReplacementStrings);
```
If correctedBoundingBox is nullopt then we should early return.  That means the frame is gone or the process has crashed, and either way we don't need to show anything.
I'd also rename correctedBoundingBox to convertedBoundingBox

achristensen07 commented on Source/WebCore/editing/AlternativeTextController.cpp:
```diff
@@ -182,8 +182,11 @@ void AlternativeTextController::show(const SimpleRange& rangeToReplace, const St
-    if (CheckedPtr client = alternativeTextClient())
-        client->showCorrectionAlternative(m_type, boundingBox, m_originalText, replacement, { });
+
+    std::optional<FrameIdentifier> rootFrameID = m_document->view()->rootFrameID();
```
Null check document and view and early return if either is null.  Same below.

stwrt commented on Source/WebKit/WebProcess/WebCoreSupport/mac/WebAlternativeTextClient.cpp:
```diff
@@ -47,9 +47,9 @@ WebAlternativeTextClient::~WebAlternativeTextClient()
+void WebAlternativeTextClient::showCorrectionAlternative(AlternativeTextType type, const FloatRect& boundingBoxOfReplacedString, const String& replacedString, const String& replacementString, const Vector<String>& alternativeReplacementStrings, std::optional<FrameIdentifier> rootFrame)
 {
-    m_page->send(Messages::WebPageProxy::ShowCorrectionPanel(type, boundingBoxOfReplacedString, replacedString, replacementString, alternativeReplacementStrings));
+    m_page->send(Messages::WebPageProxy::ShowCorrectionPanel(type, boundingBoxOfReplacedString, replacedString, replacementString, alternativeReplacementStrings, rootFrame));
```
Should be rootFrameID

achristensen07 commented on Source/WebCore/editing/AlternativeTextController.cpp:
```diff
@@ -182,8 +182,13 @@ void AlternativeTextController::show(const SimpleRange& rangeToReplace, const St
-    if (CheckedPtr client = alternativeTextClient())
-        client->showCorrectionAlternative(m_type, boundingBox, m_originalText, replacement, { });
+
+    if (!m_document.ptr() || !m_document->view())
```
m_document's type indicates it doesn't need a null check.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -13139,10 +13139,19 @@ void WebPageProxy::substitutionsPanelIsShowing(CompletionHandler<void(bool)>&& c
+        if (!convertedBoundingBox)
+            return;
+
+        pageClient->showCorrectionPanel(panelType, convertedBoundingBox.value(), replacedString, replacementString, alternativeReplacementStrings);
```
nit: I tend to like *convertedBoundingBox instead of convertedBoundingBox.value() because it makes it more clear that it is something that should've already been checked.  You do the check so it's fine either way.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -13139,10 +13139,19 @@ void WebPageProxy::substitutionsPanelIsShowing(CompletionHandler<void(bool)>&& c
+        return;
+
+    convertRectToMainFrameCoordinates(boundingBoxOfReplacedString, rootFrameID, [pageClient = WTFMove(pageClient), panelType = WTFMove(panelType), replacedString = WTFMove(replacedString),
+        replacementString = WTFMove(replacementString), alternativeReplacementStrings = WTFMove(alternativeReplacementStrings)](std::optional<WebCore::FloatRect> convertedBoundingBox) {
```
Let's put this all on the same line or give each one its own line.

stwrt committed:
[Site Isolation] Fix Spell Check
https://bugs.webkit.org/show_bug.cgi?id=292012
rdar://117912363

Reviewed by Alex Christensen.

When calculating the bounding box, the inner iframe is not able to see the outer iframes location. To correct for this missing
information, the UI process will then correct the received bounding boxes location to draw the correct UI pop up.

* Source/WebCore/editing/AlternativeTextController.cpp:
(WebCore::AlternativeTextController::show):
(WebCore::AlternativeTextController::timerFired):
* Source/WebCore/page/AlternativeTextClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::showCorrectionPanel):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebAlternativeTextClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebAlternativeTextClient.cpp:
(WebKit::WebAlternativeTextClient::showCorrectionAlternative):
* Source/WebKitLegacy/mac/WebCoreSupport/WebAlternativeTextClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebAlternativeTextClient.mm:
(WebAlternativeTextClient::showCorrectionAlternative):

Canonical link: https://commits.webkit.org/294111@main
